ex2/FunSets.scala

```diff
@@ -0,0 +1,95 @@
+package funsets
+
+
+/**
+  * 2. Purely Functional Sets.
+  */
+object FunSets {
+  /**
+    * We represent a set by its characteristic function, i.e.
+    * its `contains` predicate.
+    */
+  type Set = Int => Boolean
+
+  /**
+    * Indicates whether a set contains a given element.
+    */
+  def contains(s: Set, elem: Int): Boolean = s(elem)
+
+  /**
+    * Returns the set of the one given element.
+    */
+  def singletonSet(elem: Int): Set = { otherElem: Int => otherElem == elem }
+
+
+  /**
+    * Returns the union of the two given sets,
+    * the sets of all elements that are in either `s` or `t`.
+    */
+  def union(s: Set, t: Set): Set = { elem: Int => contains(s, elem) || contains(t, elem) }
+
+  /**
+    * Returns the intersection of the two given sets,
+    * the set of all elements that are both in `s` and `t`.
+    */
+  def intersect(s: Set, t: Set): Set = { elem: Int => contains(s, elem) && contains(t, elem) }
+
+  /**
+    * Returns the difference of the two given sets,
+    * the set of all elements of `s` that are not in `t`.
+    */
+  def diff(s: Set, t: Set): Set = { elem: Int => contains(s, elem) && !contains(t, elem) }
+
+  /**
+    * Returns the subset of `s` for which `p` holds.
+    */
+  def filter(s: Set, p: Int => Boolean): Set = { elem: Int => contains(s, elem) && p(elem) }
+
+
+  /**
+    * The bounds for `forall` and `exists` are +/- 1000.
+    */
+  val bound = 1000
+
+  /**
+    * Returns whether all bounded integers within `s` satisfy `p`.
+    */
+  def forall(s: Set, p: Int => Boolean): Boolean = {
+    def iter(a: Int): Boolean = {
+      // if we itered all elements
+      if (a > bound) true
+      // if an element doesn't conform to function p
+      else if (contains(s, a) && !p(a)) false
+      //iter next
+      else iter(a + 1)
+    }
+    //iter from initial val
+    iter(-bound)
+  }
+
+  /**
+    * Returns whether there exists a bounded integer within `s`
+    * that satisfies `p`.
+    */
+  def exists(s: Set, p: Int => Boolean): Boolean = {!forall(s, elem => !p(elem))}
```

Comment:
try not to use double negations.
maybe clearer will be to extract the `iter` method from `forall` and use it here